packages/utils/telemetry-utils/src/errorLogging.ts

```diff
@@ -0,0 +1,172 @@
+            stackFrames.unshift(errorName); // Add "[ErrorName]"
+        }
+        return stackFrames.join("\n");
+    };
+
+    const message = (typeof error?.message === "string")
+        ? error.message as string
+        : String(error);
+
+    const safeProps: { message: string; errorType?: string; stack?: string } = {
+        message,
+    };
+
+    if (isRegularObject(error)) {
+        const { errorType, stack, name } = error;
+
+        if (typeof errorType === "string") {
+            safeProps.errorType = errorType;
+        }
+
+        if (typeof stack === "string") {
+            const errorName = (typeof name === "string") ? name : undefined;
+            safeProps.stack = removeMessageFromStack(stack, errorName);
+        }
+    }
+
+    return safeProps;
+}
+
+/** Type guard for ILoggingError */
+export const isILoggingError = (x: any): x is ILoggingError => typeof x?.getTelemetryProperties === "function";
+
+/**
+ * Read-Write Logging Error.  Not exported.
+ * This type alias includes addTelemetryProperties, and applies to objects even if not instanceof LoggingError\
+ */
+type RwLoggingError = LoggingError;
+
+/** Type guard for RwLoggingError.  Not exported. */
+const isRwLoggingError = (x: any): x is RwLoggingError =>
+    typeof x?.addTelemetryProperties === "function" && isILoggingError(x);
+
+/**
+ * Annotate the given error object with the given logging props
+ * @returns The same error object passed in if possible, with telemetry props functionality mixed in
+ */
+export function annotateError(
+    error: unknown,
+    props: ITelemetryProperties,
+): ILoggingError {
+    if (isRwLoggingError(error)) {
+        error.addTelemetryProperties(props);
+        return error;
+    }
+
+    if (isRegularObject(error)) {
+        // Even though it's not exposed, fully implement IRwLoggingError for subsequent calls to annotateError
+        const loggingError = error as RwLoggingError;
+
+        const propsForError = {...props};
+        loggingError.getTelemetryProperties = () => propsForError;
+        loggingError.addTelemetryProperties =
+            (newProps: ITelemetryProperties) => { copyProps(propsForError, newProps); };
+        return loggingError;
+    }
+
+    const message = String(error);
+    return new LoggingError(message, props);
+}
+
+/** Copy props from source onto target, overwriting any keys that are already set on target */
+function copyProps(target: unknown, source: ITelemetryProperties) {
+    Object.assign(target, source);
+}
+
+/**
+ * Type guard to identify if a particular value (loosely) appears to be a tagged telemetry property
+ */
+ export function isTaggedTelemetryPropertyValue(x: any): x is ITaggedTelemetryPropertyType {
+    return (typeof(x?.value) !== "object" && typeof(x?.tag) === "string");
+}
+
+/**
+ * Walk an object's enumerable properties to find those fit for telemetry.
+ */
+function getValidTelemetryProps(obj: any): ITelemetryProperties {
```

Comment:
This tagging stuff should move elsewhere, but that won't happen until after #5560 is fully resolved.